Fix few issues on Windows

indyone opened the pull request:
Hi, @rgreenjr 

As I was trying to use the Elixir formatter on my Windows environment with Atom 1.21.1 and Elixir 1.6 (compiled from source), I encountered a few issues and I did some fixes...

1. The first was that on formatting an `.ex` file (either on save or by the menu) i got the "Elixir Formatter Error" without any details. I found that a similar issue was reported before (#2), I'm not usre if it is the same case, but in my case I had an error coming from Node.js.
The commit 73f18c5 fixes this issue.
2. Then, i got the error:

```
Error: spawnSync mix ENOENT
    at exports._errnoException (util.js:1022:11)
    at Object.spawnSync (child_process.js:480:20)
    at Object.runFormat (file:///C:/Users/indyone/.atom/packages/atom-elixir-formatter/lib/formatter.js:62:26)
    at Object.formatTextEditor (file:///C:/Users/indyone/.atom/packages/atom-elixir-formatter/lib/formatter.js:26:54)
    at file:///C:/Users/indyone/.atom/packages/atom-elixir-formatter/lib/main.js:37:19
    at Function.module.exports.Emitter.simpleDispatch (C:\Users\indyone\AppData\Local\atom\app-1.21.1\resources\app\node_modules\event-kit\lib\emitter.js:25:20)
    at Emitter.<anonymous> (C:\Users\indyone\AppData\Local\atom\app-1.21.1\resources\app\node_modules\event-kit\lib\emitter.js:154:50)
    at Emitter.module.exports.Emitter.emitAsync (C:\Users\indyone\AppData\Local\atom\app-1.21.1\resources\app\node_modules\event-kit\lib\emitter.js:157:18)
    at C:\Users\indyone\AppData\Local\atom\app-1.21.1\resources\app\node_modules\text-buffer\lib\text-buffer.js:1288:38
```
  The issue here is that on Windows, `child_process.spawn` ignores `PATHEXT`. According to the [documentation](https://nodejs.org/api/child_process.html#child_process_spawning_bat_and_cmd_files_on_windows) you can use the `{shell: true}` option to fix that. That's my 2nd commit fe16215 

3. After fixing that, i got the error "No file named mix". I found out that the command that was running for formatting the file was `elixir mix format -`, which indeed produces this error. Your commit 35027e8c17e516027047696b523a85951926d81d does not seem to work for me, so quickly reverted it (just the offending code, no tests etc) with my commit aec32ec.

After these changes, Elixir Formatter works flawlessly! 
Please check my fixes. Thanks

*Note: Unfortunately, I haven't tested this on MacOS or Linux to see if these changes break something else.*

indyone commented:
@rgreenjr If you would like to keep the fix for the 3rd issue, I can do a proper revert and make Travis to pass :)

rgreenjr commented:
Hi @indyone – these are great contributions. I'm on the road until Sunday, but will review everything then. Thanks! 👍

rgreenjr commented:
@indyone The first two commits look great. The tests in `formatter-spec.js`  just need to be updated to account for the addition of the `shell: true` option. Can you update those test so they pass?
 
I'm puzzled about the "No file named mix" error on Windows. The `elixir` and `mix` executables should both be in the `bin` directory of your Elixir 1.6 installation. Do you have the `atom-elixir-formatter` settings for **Elixir Executable** and **Mix Executable** set to the absolute paths of the `elixir` and `mix` executables?

indyone commented:
> Can you update those test so they pass?

Yes, ofc! I will do the appropriate changes for the test :) 

> I'm puzzled about the "No file named mix" error on Windows. The elixir and mix executables should both be in the bin directory of your Elixir 1.6 installation. Do you have the atom-elixir-formatter settings for Elixir Executable and Mix Executable set to the absolute paths of the elixir and mix executables?

@rgreenjr this error "No file name mix" is returned when you run the command `elixir mix format -` on Windows (https://imgur.com/a/XJyL7). It seems that on Windows is not the preferred way to run `mix` as argument of `elixir`, because the argument `mix` is not resolved to the script located in the `bin\mix` but instead it looks for a file called `mix` in the current folder.

When I use absolute paths and it works properly, eg.
* Elixir Executable: `C:\Program Files (x86)\Elixir-1.6-dev\bin\elixir` (Works also with the `elixir.bat`)
* Mix Executable: `C:\Program Files (x86)\Elixir-1.6-dev\bin\mix` (Won't work with `mix.bat`)

FYI, on Windows the executables for Elixir & Mix are not the `bin\elixir` & `bix\mix`, but are the `.bat` versions... `elixir.bat` & `mix.bat`

So, my conclusion is that the default settings won't work on Windows. If you set an absolute path to Elixir & Mix executables it will work. By changing `atom-elixir-formatter ` to call the command `mix format -`, also works (and i think it also works on Linux & MacOS).

@rgreenjr What was the initial intention for changing the code to run `elixir mix format -`?

indyone commented:
A small update...

I checked the `mix.bat` code (https://github.com/elixir-lang/elixir/blob/master/bin/mix.bat) and actually what they do is calling `elixir.bat` and `mix` with the "%~dp0" as the path which searches the directories listed in the PATH environment variable for `elixir.bat` & `mix` respectively and expands to the drive letter and path of the first one found.

rgreenjr commented:
Using `elixir mix format -` instead of `mix format -` gives users control over which version of Elixir executes the `mix` script (via the `atom-elixir-formatter` **Elixir Executable** and **Mix Executable** settings). This is necessary since people often have multiple versions of Elixir installed (using tools like `asdf`, `kiex`, and `brew`). It's also the only way to use the Elixir 1.6 code formatter on a project being developed on Elixir 1.5.2 or earlier. Without it, the first `elixir` executable resolved on the user's path will always be used.

It's interesting that the `bin\mix.bat` file on Windows differs from the `bin/mix` file. I'm not very familiar with Windows scripting. Do you know why the `elixir.bat` call is embedded in `bin\mix.bat`?

indyone commented:
> Using elixir mix format - instead of mix format - gives users control over which version of Elixir executes the mix script (via the atom-elixir-formatter Elixir Executable and Mix Executable settings). ...

OK then, I agree. We need to keep the command `elixir mix format -`.

> Do you know why the elixir.bat call is embedded in bin\mix.bat?

The reason is that the `bin\mix` is actually a Elixir script file and has the shebang `#!/usr/bin/env elixir` for telling bash (or others) to use elixir to run this script. Ofc, this does not work ok Windows cmd. 
So, here comes the `.bat` file which Windows understands and can run it.

As for the `bin\elixir.bat` is the "batch" version of the `bin\elixir` which is a Bash script file.

What if, for Windows only, when the Elixir & Mix executables have the default values in settings, we expand the paths, as `mix.bat` does? Do you think this is a good idea as a solution for this?

Another solution that comes to my mind, is to ignore the "Elixir Executable" and run the command `mix format -` (again, for Windows only and when using the default values for Elixir & Mix executables on settings).

rgreenjr commented:
Great, thanks for the clarification.

If you change **Elixir Executable** in settings to the absolute path of `elixir.bat`, will it be used to execute `mix.bat`? Or does the embedded call to `"%~dp0\elixir.bat"` in `mix.bat` get expanded and used instead? It seems like it must be the latter, which is different behavior than on Linux and macOS, and therefore the source of this bug.

indyone commented:
> If you change Elixir Executable in settings to the absolute path of elixir.bat, will it be used to execute mix.bat?

No, it won't be used. Instead, the `%~dp0\elixir.bat` will get expanded and used in `mix.bat`.

May I suggest something? 
What if, instead of having settings for the paths to `Elixir Executable` & `Mix Executable`, to have only one setting for the `Elixir Bin Path`. This way we can have code that will resolve properly the `elixir mix format -` command for Linux/MacOS and Windows, and we also support multiple installations of Elixir, and you can have just one setting to configure it!

What do you think?

indyone commented:
FYI, For those that want to run the current version of `atom-elixir-formatter` (0.2.2) on Windows, they have to set the following settings:

* **Elixir Executable**: `<path-to-elixir>\bin\elixir.bat` *(Note: Must point to the `elixir.bat` file)*
* **Mix Executable**: `<path-to-elixir>\bin\mix` *(Note: Must point to the `mix` file)*

where `<path-to-elixir>` is the Elixir installation folder.

rgreenjr commented:
I like the idea of having only one setting called **Elixir Bin Path**. But won't there still be a problem on Windows? For example, if a user set **Elixir Bin Path** to `C:\elixir\bin` and `atom-elixir-formatter` ran the command:

`C:\elixir\bin\elixir.bat C:\elixir\bin\mix.bat format -`

The `%~dp0\elixir.bat` would still be expanded and used in `mix.bat`, right? 

If not, then it seems like setting **Elixir Executable** to `C:\elixir\bin\elixir.bat` and **Mix Executable** to `C:\elixir\bin\elixir.bat` should work.

Maybe I'm missing something.

rgreenjr commented on lib/formatter.js:
```diff
@@ -23,18 +23,15 @@ export default {
+      } else {
         this.showErrorNotifcation("Elixir Formatter Error", {
-          detail: stderr
+          detail: stderr || error.message
```
Good news – the `detail` arg will have `toString()` called on it automatically, so this can be `detail: stderr || error`. This is nice since we also don't have to null check `error` before calling `message`.

rgreenjr commented on lib/formatter.js:
```diff
@@ -23,18 +23,15 @@ export default {
 
-      switch (status) {
-      case 0: {
+      if (status == 0 && !stderr.length && !error) {
```
`stderr` can null, so we have to check it first before calling `length`. Maybe this should be:

`if (status == 0 && !error) {`

indyone replied:
@rgreenjr i have updated the code so that it will handle properly the case when `stderr` is null, and also replaced `error.message` with just `error`.

indyone commented:
> I like the idea of having only one setting called Elixir Bin Path. But won't there still be a problem on Windows? For example, if a user set Elixir Bin Path to C:\elixir\bin and atom-elixir-formatter ran the command...

I'll give you a few examples of what I mean...

On Linux/MacOS if you set the **Elixir Bin Path** to `/usr/local/sbin/elixir-1.6/bin` then the command that will be produced will be:

`/usr/local/sbin/elixir-1.6/bin/elixir /usr/local/sbin/elixir-1.6/bin/mix format -`

On Windows if you set the **Elixir Bin Path** to `C:\Program Files\Elixir-1.6\bin` then the command will be:

`cmd "C:\Program Files\Elixir-1.6\bin\elixir.bat" "C:\Program Files\Elixir-1.6\bin\mix" format -`

To make this work is to have code "generates" properly the commands for Windows and the other OSes.

rgreenjr commented:
Would you mind testing something? Try setting **Elixir Executable** and **Mix Executable** to the absolute paths of your Elixir 1.6 `elixir.bat` and `mix.bat` files. Does that work?

indyone commented:
@rgreenjr i have updated the code so that it will handle properly the case when `stderr` is null, and also replaced `error.message` with just `error`.

Also for the `shell: true` option in the `spawnLink` call, i have updated the tests.

> Would you mind testing something? Try setting Elixir Executable and Mix Executable to the absolute paths of your Elixir 1.6 elixir.bat and mix.bat files. Does that work?

Yes, i tested it by setting the paths as shown in the screenshot below and I'm getting the error: 
```
    (SyntaxError) c:/Program Files/Elixir-1.6-dev/bin/mix.bat:1: syntax error before: '('
    (elixir) lib/code.ex:678: Code.require_file/2``
```
![capture](https://user-images.githubusercontent.com/1120410/32468731-289fa9e8-c358-11e7-8682-2b1a84e2793b.PNG)

On Windows, the only way to make it work (that includes the case when using the default settings), is by setting the **Elixir Executable** to `<path-to-elixir-bin>\elixir.bat` and the **Mix Executable** to `<path-to-elixir-bin>\mix`, eg.

![capture](https://user-images.githubusercontent.com/1120410/32468895-d82c8eb2-c358-11e7-87ca-9c1687de309a.PNG)

Note that you also need to encapsulate the path in double quotes if your path contains spaces.

rgreenjr commented:
Thanks @indyone! 👍 

I've merge and published a release with your changes to display node errors and fix the "spawnSync elixir ENOENT" error on Windows.

I'll open an issue and investigate the double quote problem with **Elixir Executable** and *Mix Executable** settings. Users shouldn't have to worry quoting those.

indyone commented:
@rgreenjr thank you too.

I'll open an additional issue, that when using the default settings on Windows, on format, you get the error `No file named mix`, as it is not still resolve with this PR.
